cpp/oneapi/dal/algo/logistic_regression/backend/gpu/infer_kernel_sparse_dpc.cpp

```diff
@@ -0,0 +1,140 @@
+#include "oneapi/dal/backend/primitives/ndindexer.hpp"
+
+#include "oneapi/dal/table/csr_accessor.hpp"
+
+#include "oneapi/dal/algo/logistic_regression/common.hpp"
+#include "oneapi/dal/algo/logistic_regression/train_types.hpp"
+#include "oneapi/dal/algo/logistic_regression/backend/model_impl.hpp"
+#include "oneapi/dal/algo/logistic_regression/backend/gpu/infer_kernel.hpp"
+#include "oneapi/dal/backend/primitives/sparse_blas.hpp"
+
+namespace oneapi::dal::logistic_regression::backend {
+
+using dal::backend::context_gpu;
+
+namespace be = dal::backend;
+namespace pr = be::primitives;
+namespace interop = dal::backend::interop;
+
+template <typename Float, typename Task>
+static infer_result<Task> call_dal_kernel(const context_gpu& ctx,
+                                          const detail::descriptor_base<Task>& desc,
+                                          const table& infer,
+                                          const model<Task>& m) {
+    auto queue = ctx.get_queue();
+    ONEDAL_PROFILER_TASK(logreg_infer_kernel, queue);
+
+    constexpr auto alloc = sycl::usm::alloc::device;
+
+    const auto& betas = m.get_packed_coefficients();
+
+    const auto sample_count = infer.get_row_count();
+    const auto feature_count = infer.get_column_count();
+    const bool fit_intercept = desc.get_compute_intercept();
+    ONEDAL_ASSERT((feature_count + 1) == betas.get_column_count());
+    ONEDAL_ASSERT(1 == betas.get_row_count());
+    ONEDAL_ASSERT(infer.get_kind() == dal::csr_table::kind());
+
+    pr::ndarray<Float, 1> params = pr::table2ndarray_1d<Float>(queue, betas, alloc);
+    pr::ndview<Float, 1> params_suf = fit_intercept ? params : params.slice(1, feature_count);
+
+    pr::ndarray<Float, 1> probs = pr::ndarray<Float, 1>::empty(queue, { sample_count }, alloc);
+    pr::ndarray<std::int32_t, 1> responses =
+        pr::ndarray<std::int32_t, 1>::empty(queue, { sample_count }, alloc);
+
+    auto [csr_data, column_indices, row_offsets] =
+        csr_accessor<const Float>(static_cast<const csr_table&>(infer))
+            .pull(queue, { 0, -1 }, sparse_indexing::zero_based);
+
+    auto csr_data_gpu =
+        pr::ndarray<Float, 1>::wrap(csr_data.get_data(), csr_data.get_count()).to_device(queue);
+    auto column_indices_gpu =
+        pr::ndarray<std::int64_t, 1>::wrap(column_indices.get_data(), column_indices.get_count())
+            .to_device(queue);
+    auto row_offsets_gpu =
+        pr::ndarray<std::int64_t, 1>::wrap(row_offsets.get_data(), row_offsets.get_count())
+            .to_device(queue);
+
+    table data_gpu = csr_table::wrap(queue,
+                                     csr_data_gpu.get_data(),
+                                     column_indices_gpu.get_data(),
+                                     row_offsets_gpu.get_data(),
+                                     sample_count,
+                                     feature_count,
+                                     sparse_indexing::zero_based);
```

Comment:
This code just moves the data from any USM storage to device.
Please move this into a separate function. Also it seems we do not need to transfer the data to device if it's already on device.

Reply:
We would need to create a separate class for this so that csr_data_gpu is not deleted after exiting the function, so it's that easy. I suggest doing this in a separate pull request so that we could use it in other algorithms